### PR TITLE
Support auto-closing quotes in Python raw string literals, etc

### PR DESCRIPTION
**Pull request by Syeberman**
Python uses a ["string prefix"](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals) to denote bytes literals (`b''`), raw strings (`r''`), and formatted string literals (`f''`).  This change makes it so that when one types `r'` in VS Code, the string will be auto-closed as `r''` (as an example).

Python also supports multiple string prefixes for a single literal.  I've tested that typing `fr'` will auto-close to `fr''`, and so forth.

### CODE REVIEW

File: extensions/python/language-configuration.json [hunk 1]
```diff
@@ -13,7 +13,23 @@
...
+		{ "open": "B\"", "close": "\"", "notIn": ["string", "comment"] },
+		{ "open": "'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "r'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "R'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "u'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "U'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "f'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "F'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "b'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "B'", "close": "'", "notIn": ["string", "comment"] }
```
Comment (javadev) [inline extensions/python/language-configuration.json]: It may be unit tested.

**Comment by aeschli** [comment]
Isn't it enough to just have `'` and '"' as an auto pair character? The only difference I'm aware of is that undo would only remove the quote, instead of the letter before, but IMO that's better anyway.

**Comment by Syeberman** [comment]
The original auto-close setting did not auto-close `r'`, etc.  Presumably, it's coded to only trigger after a whitespace (or non-word boundary), but I didn't look at the code.

Oddly, this doesn't explain why the above configuration auto-closes `fr'`, etc.  If the open pattern must be preceded by whitespace, then that shouldn't work.  Perhaps it's coded differently if the pattern starts with an alphanumeric character?

**Comment by aeschli** [comment]
Yes, there's a [condition](https://github.com/Microsoft/vscode/blob/1ab63ddad5d7a390c337bfdc51cb6a82918a70c9/src/vs/editor/common/controller/cursorTypeOperations.ts#L499) to not autoclose `'` after a word.
@alexandrudima Do you want to reconsider that, or is the PR ok?

**Comment by aeschli** [comment]
Sorry, this slipped. @alexandrudima Waiting for your opinion.

**Comment by alexandrudima** [comment]
The following added rules do nothing:

```
		{ "open": "r\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "R\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "u\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "U\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "f\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "F\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "b\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "B\"", "close": "\"", "notIn": ["string", "comment"] },
		{ "open": "r'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "R'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "u'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "U'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "f'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "F'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "b'", "close": "'", "notIn": ["string", "comment"] },
		{ "open": "B'", "close": "'", "notIn": ["string", "comment"] }
```

Out of the added rules, this is the only one that might do something:
```
		{ "open": "'", "close": "'", "notIn": ["string", "comment"] },
```

Closing Pairs are implemented to be working on single characters.

**Comment by Syeberman** [comment]
@alexandrudima I tested this configuration myself when I submitted this PR.  Without my changes, when you type `r'`, it does not auto-close to `r''`.  With my changes, it properly auto-closes raw, unicode, f-strings, and bytes.

**Comment by alexandrudima** [comment]
Yes, that's what I'm saying too. The only line that makes this work from this PR is:
```
{ "open": "'", "close": "'", "notIn": ["string", "comment"] },
```
This defines that `'` - `'` forms an auto-closing pair.

The other 16 lines serve no purpose.

**Comment by aeschli** [comment]
@alexandrudima 
`{ "open": "'", "close": "'", "notIn": ["string", "comment"] },` is already part of the autoClosingPair.
It doesn't work after `r'` due to [this](https://github.com/Microsoft/vscode/blob/1ab63ddad5d7a390c337bfdc51cb6a82918a70c9/src/vs/editor/common/controller/cursorTypeOperations.ts#L499).
Adding `{ "open": "r'", "close": "'", "notIn": ["string", "comment"] },` fixes the problem

**Comment by alexandrudima** [comment]
@Syeberman @aeschli I agree the change makes this feature work. I was quite stumped at first, since I am pretty sure how auto-closing characters should work and how they're implemented. 

As I thought, the execution does not enter the auto closing open character code (since that is implemented to only work on single characters), but it enters `BracketElectricCharacterSupport._onElectricAutoClose` and these newly added auto-closing pairs somehow make their way to `_complexAutoClosePairs`.

@aeschli Since this is something you've authored in ebc30748223067bec60b7bac0c3becff77d27e77 , I cannot be of further assistance. 

Perhaps now would be a good time to document somewhere this behaviour!

**Comment by aeschli** [comment]
Talked to @alexandrudima again and we are fine with the PR. Not autoclosing `'` and `"` after a letter is the expected behaviour most languages, so the PR makes sense to add additional rules fo the cases where this makes sense.

Thanks @Syeberman!

**Comment by diazgilberto** [comment]
I still have the issue on `Version 1.21.1 (1.21.1)` and running OSX Sierra.

**Comment by aeschli** [comment]
@diazgilberto Please file a new issue with a code sample that shows the problem. Also try to run without extensions to see whether an external extension causes this.

**Pull request merged**